core/server/master/src/main/java/alluxio/master/job/FileIterable.java

```diff
@@ -151,28 +152,37 @@ private void partialListFileInfos() {
       if (!mStartAfter.isEmpty()) {
         mListOptions.setDisableAreDescendantsLoadedCheck(true);
       }
-      ListStatusContext context = ListStatusContext.create(ListStatusPartialPOptions
-          .newBuilder()
-          .setOptions(mListOptions)
-          .setBatchSize(PARTIAL_LISTING_BATCH_SIZE)
-          .setStartAfter(mStartAfter));
-      listFileInfos(context);
+      Supplier<ListStatusContext> context = () -> {
+        return ListStatusContext.create(ListStatusPartialPOptions
+            .newBuilder()
+            .setOptions(mListOptions)
+            .setBatchSize(PARTIAL_LISTING_BATCH_SIZE)
+            .setStartAfter(mStartAfter));
+      };
+
+      List<FileInfo> fileInfos;
+      while ((fileInfos = listStatus(context.get())) != null
```

Comment:
in which situation this would be null? doesn't seems like we would get null

Reply:
Condition ` !fileInfos.isEmpty()` should be judged after condition `(mFiles = fileInfos.stream().filter(mFilter).collect(Collectors.toList())).isEmpty()`, otherwise `mFiles` may not be updated. I added a judgment of whether or not it is null to update fileInfos first.